File: src/agent-evaluator.js
/**
 * AGENT EVALUATOR - Independent Quality Assessment Layer
 *
 * Purpose: Evaluate quality of MAP agent outputs independently
 * Can be invoked directly via API with Anthropic token
 * Not dependent on the frontend application
 *
 * Usage:
 * ```
 * const evaluator = new AgentEvaluator(apiKey);
 * const result = await evaluator.evaluate(agentName, output, context);
 * ```
 */

const Anthropic = require("@anthropic-ai/sdk");

class AgentEvaluator {
  constructor(apiKey) {
    this.client = new Anthropic({ apiKey });
    // Scoring against a fixed, structured rubric is mechanical enough for
    // haiku — this can run on every single agent invocation, so it's the
    // highest-volume call in the system. Bump back to sonnet if scores start
    // looking unreliable in practice.
    this.model = "claude-3-5-haiku-20241022";

    this.evaluationCriteria = {
      gimena: {
        name: "Gimena - User Story Writer",
        dimensions: {
          format_compliance: {
            weight: 0.2,
            description: "HU format compliance (As/I want/For structure)"
          },
          completeness: {
            weight: 0.3,
            description: "Complete AC, error handling, visual references"
          },
          clarity: {
            weight: 0.2,
            description: "Language clarity, technical accuracy"
          },
          actionability: {
            weight: 0.2,
            description: "Clear next steps, no ambiguity"
          },
          alignment: {
            weight: 0.1,
            description: "Business & technical requirements met"
          }
        },
        targetScore: 90
      },

      gabi: {
        name: "Gabi - Work Planner",
        dimensions: {
          phase_completeness: {
            weight: 0.2,
            description: "All 5 phases detailed (DB→Backend→API→Frontend→Integration)"
          },
          tdd_strategy: {
            weight: 0.25,
            description: "Unit, Integration, API, E2E test strategies"
          },
          solid_application: {
            weight: 0.2,
            description: "All 5 SOLID principles documented"
          },
          estimations: {
            weight: 0.2,
            description: "Optimistic, Realistic, Pessimistic with confidence"
          },
          risk_management: {
            weight: 0.15,
            description: "Risk assessment with mitigation"
          }
        },
        targetScore: 85
      },

      gabriela: {
        name: "Gabriela - Project Brain",
        dimensions: {
          stakeholder_clarity: {
            weight: 0.2,
            description: "Complete stakeholder table with roles"
          },
          scope_definition: {
            weight: 0.25,
            description: "In/Out scope matrix clarity"
          },
          timeline_completeness: {
            weight: 0.15,
            description: "Start, end, milestones defined"
          },
          business_rules: {
            weight: 0.2,
            description: "Critical rules documented"
          },
          documentation_currency: {
            weight: 0.2,
            description: "Meeting log and change log up-to-date"
          }
        },
        targetScore: 90
      },

      santi: {
        name: "Santi - Meeting Minutes",
        dimensions: {
          format_compliance: {
            weight: 0.25,
            description: "H1/H2/H3 hierarchy correct (no H2 numbers)"
          },
          action_items: {
            weight: 0.25,
            description: "☐ symbol usage, owner, deadline"
          },
          language_precision: {
            weight: 0.2,
            description: "Direct language, no 'mentioned that' phrases"
          },
          risk_identification: {
            weight: 0.2,
            description: "Alerts and red flags clearly marked"
          },
          professional_tone: {
            weight: 0.1,
            description: "Executive, results-oriented tone"
          }
        },
        targetScore: 85
      },

      daniel: {
        name: "Daniel - Release Notes",
        dimensions: {
          technical_accuracy: {
            weight: 0.25,
            description: "Bitbucket version accuracy"
          },
          dual_version_coherence: {
            weight: 0.25,
            description: "Bitbucket ↔ Basecamp alignment"
          },
          audience_appropriateness: {
            weight: 0.2,
            description: "Technical vs client-friendly tone"
          },
          completeness: {
            weight: 0.2,
            description: "All commits, features, fixes included"
          },
          presentation: {
            weight: 0.1,
            description: "Professional, clear structure"
          }
        },
        targetScore: 90
      }
    };
  }

  /**
   * Main evaluation method
   */
  async evaluate(agentName, output, context = {}) {
    const criteria = this.evaluationCriteria[agentName] || this.genericCriteria(agentName);
    const prompt = this.buildEvaluationPrompt(agentName, criteria, output, context);

    try {
      const response = await this.client.messages.create({
        model: this.model,
        max_tokens: 2000,
        messages: [
          {
            role: "user",
            content: prompt
          }
        ]
      });

      const evaluationText = response.content[0].text;
      const scores = this.parseEvaluationResponse(evaluationText, criteria);

      return {
        agent: agentName,
        timestamp: new Date().toISOString(),
        scores: scores,
        targetScore: criteria.targetScore,
        status: this.getStatus(scores.overall, criteria.targetScore),
        feedback: evaluationText,
        dimensions: criteria.dimensions
      };
    } catch (error) {
      throw new Error(`Evaluation failed for ${agentName}: ${error.message}`);
    }
  }

  /**
   * Fallback criteria for agents without a bespoke rubric (e.g. the 14
   * spec-kit agents) so /evaluate never 400s on a known, invokable agent.
   */
  genericCriteria(agentName) {
    return {
      name: agentName,
      dimensions: {
        completeness: { weight: 0.3, description: "Output covers what was asked" },
        clarity: { weight: 0.25, description: "Clear, unambiguous" },
        adherence: { weight: 0.25, description: "Follows the agent's documented role/restrictions" },
        actionability: { weight: 0.2, description: "Next steps are clear" }
      },
      targetScore: 80
    };
  }

  /**
   * Batch evaluate multiple outputs
   */
  async evaluateBatch(evaluations) {
    const results = [];
    for (const item of evaluations) {
      const result = await this.evaluate(item.agent, item.output, item.context);
      results.push(result);
    }
    return results;
  }

  /**
   * Build evaluation prompt
   */
  buildEvaluationPrompt(agentName, criteria, output, context) {
    const dimensionsStr = Object.entries(criteria.dimensions)
      .map(
        ([key, value]) =>
          `- ${key.replace(/_/g, " ").toUpperCase()}: ${value.description} (Weight: ${value.weight * 100}%)`
      )
      .join("\n");

    return `You are a quality assurance specialist evaluating AI agent outputs for a project management system.

AGENT: ${criteria.name}
TARGET SCORE: ${criteria.targetScore}/100

EVALUATION DIMENSIONS:
${dimensionsStr}

CONTEXT:
${JSON.stringify(context, null, 2)}

OUTPUT TO EVALUATE:
${output}

EVALUATION TASK:
1. Score each dimension from 0-100 based on the criteria
2. Calculate weighted overall score
3. Identify red flags (critical issues)
4. Provide improvement recommendations
5. Determine if output meets quality standards

RESPONSE FORMAT (CRITICAL - MUST BE VALID JSON):
{
  "dimensions": {
    "dimension_name": {
      "score": <number 0-100>,
      "reasoning": "<explanation>"
    }
  },
  "overall": <weighted average>,
  "status": "<EXCELLENT|GOOD|WARNING|CRITICAL>",
  "redFlags": ["<flag1>", "<flag2>"],
  "recommendations": ["<rec1>", "<rec2>"],
  "summary": "<brief overall assessment>"
}`;
  }

  /**
   * Parse evaluation response
   */
  parseEvaluationResponse(text, criteria) {
    try {
      const jsonMatch = text.match(/\{[\s\S]*\}/);
      if (!jsonMatch) {
        throw new Error("No JSON found in response");
      }

      const evaluation = JSON.parse(jsonMatch[0]);

      let overall = evaluation.overall || 0;
      if (overall === 0) {
        let weightedScore = 0;
        let totalWeight = 0;

        for (const [dimension, weight] of Object.entries(criteria.dimensions)) {
          const score = evaluation.dimensions[dimension]?.score || 0;
          weightedScore += score * weight.weight;
          totalWeight += weight.weight;
        }

        overall = Math.round(weightedScore / totalWeight);
      }

      return {
        dimensions: evaluation.dimensions || {},
        overall: overall,
        status: evaluation.status || this.getStatus(overall, criteria.targetScore),
        redFlags: evaluation.redFlags || [],
        recommendations: evaluation.recommendations || [],
        summary: evaluation.summary || ""
      };
    } catch (error) {
      console.error("Parse error:", error);
      return {
        dimensions: {},
        overall: 0,
        status: "CRITICAL",
        redFlags: ["Evaluation parsing failed"],
        recommendations: ["Review agent output format"],
        summary: "Unable to parse evaluation"
      };
    }
  }

  /**
   * Determine status based on score
   */
  getStatus(score, targetScore) {
    if (score >= targetScore) return "EXCELLENT";
    if (score >= 75) return "GOOD";
    if (score >= 50) return "WARNING";
    return "CRITICAL";
  }

  /**
   * Generate report from multiple evaluations
   */
  generateReport(evaluations) {
    const report = {
      timestamp: new Date().toISOString(),
      totalEvaluations: evaluations.length,
      agents: {},
      overallMetrics: {
        averageScore: 0,
        excellentCount: 0,
        goodCount: 0,
        warningCount: 0,
        criticalCount: 0
      },
      trends: [],
      recommendations: []
    };

    let totalScore = 0;

    for (const item of evaluations) {
      totalScore += item.scores.overall;

      if (!report.agents[item.agent]) {
        report.agents[item.agent] = {
          latestScore: item.scores.overall,
          targetScore: item.targetScore,
          status: item.status,
          redFlags: item.scores.redFlags,
          recommendations: item.scores.recommendations
        };
      }

      if (item.status === "EXCELLENT") report.overallMetrics.excellentCount++;
      if (item.status === "GOOD") report.overallMetrics.goodCount++;
      if (item.status === "WARNING") report.overallMetrics.warningCount++;
      if (item.status === "CRITICAL") report.overallMetrics.criticalCount++;
    }

    report.overallMetrics.averageScore = evaluations.length
      ? Math.round(totalScore / evaluations.length)
      : 0;

    return report;
  }
}

module.exports = AgentEvaluator;

File: app/services/agent_evaluator.py
"""
Agent Evaluator — migrated from src/agent-evaluator.js.

Independent quality assessment layer: scores an agent's output against a
per-agent weighted rubric (or a generic fallback for agents without one),
via a single Anthropic call on haiku (mechanical enough, and this is the
highest-volume call in the system if it runs on every invocation).
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from anthropic import AsyncAnthropic

EVALUATOR_MODEL = "claude-sonnet-4-6"


@dataclass(frozen=True)
class Dimension:
    weight: float
    description: str


@dataclass(frozen=True)
class AgentCriteria:
    name: str
    dimensions: dict[str, Dimension]
    target_score: int


EVALUATION_CRITERIA: dict[str, AgentCriteria] = {
    "gime": AgentCriteria(
        name="Gime - User Story Writer",
        dimensions={
            "format_compliance": Dimension(0.2, "HU format compliance (As/I want/For structure)"),
            "completeness": Dimension(0.3, "Complete AC, error handling, visual references"),
            "clarity": Dimension(0.2, "Language clarity, technical accuracy"),
            "actionability": Dimension(0.2, "Clear next steps, no ambiguity"),
            "alignment": Dimension(0.1, "Business & technical requirements met"),
        },
        target_score=90,
    ),
    "gabi": AgentCriteria(
        name="Gabi - Work Planner",
        dimensions={
            "phase_completeness": Dimension(0.2, "All 5 phases detailed (DB→Backend→API→Frontend→Integration)"),
            "tdd_strategy": Dimension(0.25, "Unit, Integration, API, E2E test strategies"),
            "solid_application": Dimension(0.2, "All 5 SOLID principles documented"),
            "estimations": Dimension(0.2, "Optimistic, Realistic, Pessimistic with confidence"),
            "risk_management": Dimension(0.15, "Risk assessment with mitigation"),
        },
        target_score=85,
    ),
    "gaby": AgentCriteria(
        name="Gaby - Project Brain",
        dimensions={
            "stakeholder_clarity": Dimension(0.2, "Complete stakeholder table with roles"),
            "scope_definition": Dimension(0.25, "In/Out scope matrix clarity"),
            "timeline_completeness": Dimension(0.15, "Start, end, milestones defined"),
            "business_rules": Dimension(0.2, "Critical rules documented"),
            "documentation_currency": Dimension(0.2, "Meeting log and change log up-to-date"),
        },
        target_score=90,
    ),
    "santi": AgentCriteria(
        name="Santi - Meeting Minutes",
        dimensions={
            "format_compliance": Dimension(0.25, "H1/H2/H3 hierarchy correct (no H2 numbers)"),
            "action_items": Dimension(0.25, "☐ symbol usage, owner, deadline"),
            "language_precision": Dimension(0.2, "Direct language, no 'mentioned that' phrases"),
            "risk_identification": Dimension(0.2, "Alerts and red flags clearly marked"),
            "professional_tone": Dimension(0.1, "Executive, results-oriented tone"),
        },
        target_score=85,
    ),
    "dani": AgentCriteria(
        name="Dani - Release Notes",
        dimensions={
            "technical_accuracy": Dimension(0.25, "Bitbucket version accuracy"),
            "dual_version_coherence": Dimension(0.25, "Bitbucket ↔ Basecamp alignment"),
            "audience_appropriateness": Dimension(0.2, "Technical vs client-friendly tone"),
            "completeness": Dimension(0.2, "All commits, features, fixes included"),
            "presentation": Dimension(0.1, "Professional, clear structure"),
        },
        target_score=90,
    ),
}


def generic_criteria(agent_name: str) -> AgentCriteria:
    """Fallback rubric for agents without a bespoke one (e.g. the 14 spec-kit
    agents) so /evaluate never 400s on a known, invokable agent."""
    return AgentCriteria(
        name=agent_name,
        dimensions={
            "completeness": Dimension(0.3, "Output covers what was asked"),
            "clarity": Dimension(0.25, "Clear, unambiguous"),
            "adherence": Dimension(0.25, "Follows the agent's documented role/restrictions"),
            "actionability": Dimension(0.2, "Next steps are clear"),
        },
        target_score=80,
    )


def _get_status(score: float, target_score: int) -> str:
    if score >= target_score:
        return "EXCELLENT"
    if score >= 75:
        return "GOOD"
    if score >= 50:
        return "WARNING"
    return "CRITICAL"


def _build_evaluation_prompt(criteria: AgentCriteria, output: str, context: dict[str, Any]) -> str:
    dimensions_str = "\n".join(
        f"- {key.replace('_', ' ').upper()}: {dim.description} (Weight: {dim.weight * 100}%)"
        for key, dim in criteria.dimensions.items()
    )

    return f"""You are a quality assurance specialist evaluating AI agent outputs for a project management system.

AGENT: {criteria.name}
TARGET SCORE: {criteria.target_score}/100

EVALUATION DIMENSIONS:
{dimensions_str}

CONTEXT:
{json.dumps(context, indent=2, ensure_ascii=False)}

OUTPUT TO EVALUATE:
{output}

EVALUATION TASK:
1. Score each dimension from 0-100 based on the criteria
2. Calculate weighted overall score
3. Identify red flags (critical issues)
4. Provide improvement recommendations
5. Determine if output meets quality standards

RESPONSE FORMAT (CRITICAL - MUST BE VALID JSON):
{{
  "dimensions": {{
    "dimension_name": {{
      "score": <number 0-100>,
      "reasoning": "<explanation>"
    }}
  }},
  "overall": <weighted average>,
  "status": "<EXCELLENT|GOOD|WARNING|CRITICAL>",
  "redFlags": ["<flag1>", "<flag2>"],
  "recommendations": ["<rec1>", "<rec2>"],
  "summary": "<brief overall assessment>"
}}"""


def _parse_evaluation_response(text: str, criteria: AgentCriteria) -> dict[str, Any]:
    try:
        match = re.search(r"\{[\s\S]*\}", text)
        if not match:
            raise ValueError("No JSON found in response")

        evaluation = json.loads(match.group(0))

        overall = evaluation.get("overall") or 0
        if not overall:
            weighted_score = 0.0
            total_weight = 0.0
            for dim_name, dim in criteria.dimensions.items():
                score = (evaluation.get("dimensions", {}).get(dim_name) or {}).get("score", 0)
                weighted_score += score * dim.weight
                total_weight += dim.weight
            overall = round(weighted_score / total_weight) if total_weight else 0

        return {
            "dimensions": evaluation.get("dimensions", {}),
            "overall": overall,
            "status": evaluation.get("status") or _get_status(overall, criteria.target_score),
            "redFlags": evaluation.get("redFlags", []),
            "recommendations": evaluation.get("recommendations", []),
            "summary": evaluation.get("summary", ""),
        }
    except (ValueError, json.JSONDecodeError) as error:
        print(f"agent_evaluator: parse error: {error}")
        return {
            "dimensions": {},
            "overall": 0,
            "status": "CRITICAL",
            "redFlags": ["Evaluation parsing failed"],
            "recommendations": ["Review agent output format"],
            "summary": "Unable to parse evaluation",
        }


@dataclass
class AgentEvaluator:
    """Async port of the JS AgentEvaluator class. Construct with the
    Anthropic API key; `evaluate()` does one model call per output."""

    api_key: str
    model: str = EVALUATOR_MODEL
    base_url: str | None = None
    _client: AsyncAnthropic = field(init=False, repr=False)

    def __post_init__(self) -> None:
        kwargs: dict[str, Any] = {"api_key": self.api_key}
        if self.base_url:
            kwargs["base_url"] = self.base_url
        self._client = AsyncAnthropic(**kwargs)

    async def evaluate(
        self, agent_name: str, output: str, context: dict[str, Any] | None = None
    ) -> dict[str, Any]:
        context = context or {}
        criteria = EVALUATION_CRITERIA.get(agent_name) or generic_criteria(agent_name)
        prompt = _build_evaluation_prompt(criteria, output, context)

        try:
            response = await self._client.messages.create(
                model=self.model,
                max_tokens=2000,
                messages=[{"role": "user", "content": prompt}],
            )
            evaluation_text = response.content[0].text
            scores = _parse_evaluation_response(evaluation_text, criteria)

            return {
                "agent": agent_name,
                "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
                "scores": scores,
                "targetScore": criteria.target_score,
                "status": _get_status(scores["overall"], criteria.target_score),
                "feedback": evaluation_text,
                "dimensions": {k: {"weight": v.weight, "description": v.description} for k, v in criteria.dimensions.items()},
            }
        except Exception as error:  # noqa: BLE001 - surfaced as a wrapped error, matching JS behavior
            raise RuntimeError(f"Evaluation failed for {agent_name}: {error}") from error

    async def evaluate_batch(self, evaluations: list[dict[str, Any]]) -> list[dict[str, Any]]:
        results = []
        for item in evaluations:
            result = await self.evaluate(item["agent"], item["output"], item.get("context"))
            results.append(result)
        return results

    def generate_report(self, evaluations: list[dict[str, Any]]) -> dict[str, Any]:
        report: dict[str, Any] = {
            "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            "totalEvaluations": len(evaluations),
            "agents": {},
            "overallMetrics": {
                "averageScore": 0,
                "excellentCount": 0,
                "goodCount": 0,
                "warningCount": 0,
                "criticalCount": 0,
            },
            "trends": [],
            "recommendations": [],
        }

        total_score = 0
        for item in evaluations:
            total_score += item["scores"]["overall"]

            if item["agent"] not in report["agents"]:
                report["agents"][item["agent"]] = {
                    "latestScore": item["scores"]["overall"],
                    "targetScore": item["targetScore"],
                    "status": item["status"],
                    "redFlags": item["scores"]["redFlags"],
                    "recommendations": item["scores"]["recommendations"],
                }

            status = item["status"]
            if status == "EXCELLENT":
                report["overallMetrics"]["excellentCount"] += 1
            elif status == "GOOD":
                report["overallMetrics"]["goodCount"] += 1
            elif status == "WARNING":
                report["overallMetrics"]["warningCount"] += 1
            elif status == "CRITICAL":
                report["overallMetrics"]["criticalCount"] += 1

        report["overallMetrics"]["averageScore"] = round(total_score / len(evaluations)) if evaluations else 0
        return report
